### PR TITLE
Support combined textures for Metal target

### LINKED ISSUE #4025: Metal: texture intrinsics.

### PR DESCRIPTION
**Pull request by jkwak-work**
Closes #4025

This commit is to support the combined texture cases where the sampler is not given for the texture fetching. Internally it creates a default sampler and use it for the fetching. The logic is mostly same to how it works for HLSL.

It turned out that we were missing an offset variant of `Sample()` function. This commit adds it.

This commit also fixes the cube texture related testing problems in `tests/metal/texture.slang`.

### CODE REVIEW

File: tests/metal/texture-sampler-less.slang [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+// "$1" refers to a sampler not the first parameter of the function.
+// This causes to emit an invalid code when the implementation is shared
+// between the combined and not-combined.
+// The following code,
+//   void GetDimensions(out float width) { __intrinsic_asm "*$1 = $0.get_width(0)"; }
+// gets emitted as following,
+//   (*(kernelContext_0->t1D_sampler_0) = (kernelContext_0->t1D_texture_0).get_width(0));
+// when it is expected to be,
+//   (*&width = (kernelContext_0->t1D_texture_0).get_width(0));
+//#define TEST_WHEN_DEFAULT_SAMPLER_WORKS_PROPERLY
```
Comment (jkwak-work) [inline tests/metal/texture-sampler-less.slang:15]: I am going to file a new issue for this tomorrow.

**Review by csyonghe — APPROVED** [review]
Looks good!

File: tests/metal/texture-sampler-less.slang [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+
+//TEST_INPUT: ubuffer(data=[0], stride=4):out,name outputBuffer
+RWStructuredBuffer<int> outputBuffer;
+
+// TODO: `SamplerXD` cannot be setup with TEST_INPUT method.
+
+Sampler1D<float> t1D;
```
Comment (csyonghe) [inline tests/metal/texture-sampler-less.slang:22]: We should also check to make sure both the texture and the sampler get correct binding index in the output metal source.

Comment (jkwak-work) [reply tests/metal/texture-sampler-less.slang:22]: I will try that tomorrow.

**Pull request merged**